nexus/db-queries/src/db/datastore/instance.rs

```diff
@@ -1141,75 +1159,110 @@ impl DataStore {
+                    let expected_state =
+                        api::external::DiskState::Attached(authz_instance.id());
+
+                    let attached_disk: Option<Uuid> = disk_dsl::disk
+                        .filter(disk_dsl::id.eq(disk_id))
+                        .filter(
+                            disk_dsl::attach_instance_id
+                                .eq(authz_instance.id()),
+                        )
+                        .filter(disk_dsl::disk_state.eq(expected_state.label()))
+                        .select(disk_dsl::id)
+                        .first_async::<Uuid>(conn)
+                        .await
+                        .optional()?;
 
-        if let Some(disk_id) = boot_disk_id {
-            // Ensure the disk is currently attached before updating
-            // the database.
-            let expected_state =
-                api::external::DiskState::Attached(authz_instance.id());
-
-            let attached_disk: Option<Uuid> = disk_dsl::disk
-                .filter(disk_dsl::id.eq(disk_id))
-                .filter(disk_dsl::attach_instance_id.eq(authz_instance.id()))
-                .filter(disk_dsl::disk_state.eq(expected_state.label()))
-                .select(disk_dsl::id)
-                .first_async::<Uuid>(conn)
-                .await
-                .optional()?;
+                    if attached_disk.is_none() {
+                        return Err(err.bail(Error::conflict(
+                            "boot disk must be attached",
+                        )));
+                    }
+                }
 
-            if attached_disk.is_none() {
-                return Err(
-                    err.bail(Error::conflict("boot disk must be attached"))
-                );
+                Ok(())
             }
         }
-        //
-        // NOTE: from this point forward it is OK if we update the
-        // instance's `boot_disk_id` column with the updated value
-        // again. It will have already been assigned with constraint
-        // checking performed above, so updates will just be
-        // repetitive, not harmful.
+    }
+
+    /// Set an instance's CPU count and memory size to the provided values,
+    /// within an existing transaction.
+    async fn instance_set_size_on_conn(
+        &self,
+        conn: &async_bb8_diesel::Connection<DbConnection>,
+        err: &OptionalError<Error>,
+        authz_instance: &authz::Instance,
+        ncpus: InstanceCpuCount,
+        memory: ByteCount,
+    ) -> Result<(), diesel::result::Error> {
+        use db::schema::instance::dsl as instance_dsl;
+
+        // Do not allow setting sizes of running instances to ensure that if an
+        // instance is running, its resource usage matches what we record in the
+        // database.
+        const OK_TO_SET_SIZE_STATES: &'static [InstanceState] = &[
+            InstanceState::NoVmm,
+            InstanceState::Failed,
+            InstanceState::Creating,
+        ];
```

Comment:
It occurs to me that `OK_TO_SET_SIZE_STATES` and `OK_TO_SET_BOOT_DISK_STATES` are the same list of states, and what we really want in both cases is "states in which there is no VMM actively incarnating this instance". I kind of wonder if, rather than duplicating in both the instance-resize and instance-set-boot-disk queries, we should just have a constant on `InstanceState` that's like `NOT_INCARNATED_STATES` or similar, so that these queries and any future ones can all use the same set of states, and if we add new states that have "not incarnated" semantics, we can add them to the one definition of that list of states, instead.

Reply:
on one hand i agree (and enough to go just do the thing, there's now `InstanceState::NO_VMM_STATES`\*), but from the conversations we've had we've been _pretty close_ to having caveouts for specific not-incarnated states and fields. namely "is it OK to set the boot disk in `Creating`", and does allowing that allow correctness issues (no) or a possibly-confusing-but-correct API result (yes). otoh "actions we can take on not-incarnated instances" is a much easier concept than having to think through each kind of action and state combination, so it's nicer in an axiomatic way.

\*: after writing this i realize why you said `NOT_INCARNATED` rather than `NO_VMM` - because there _is_ a `NoVmm` state which is very distinct from what this list is getting at. so i should go change that name.

Reply:
Yeah, my thinking is that we should avoid having weird exceptions for setting specific parameters unless it's absolutely necessary. I'd be quite surprised if we don't basically just end up being able to divide every possible `instance_reconfigure` into either "can do this only if there is no Real Life VMM Process incarnating the instance" and "can do this whenever", and I'd prefer to have to burn that bridge only once we come to it...